query/src/sessions/session_ref.rs

```diff
@@ -56,8 +55,7 @@ impl Drop for SessionRef {
 
 impl Session {
     pub fn destroy_session_ref(self: &Arc<Self>) {
-        if self.ref_count.fetch_sub(1, Ordering::Release) == 1 {
-            std::sync::atomic::fence(Acquire);
+        if self.ref_count.fetch_sub(1, Ordering::Relaxed) == 1 {
```

Comment:
why takes Relaxed memory ordering? in my understanding this means only modify the variable locally without memory barrier at all

Reply:
I see, dereferencing self have a memory barrier here implicitly.

but dereferencing self may not necessarily means the varaible ref_count is synced into other cores.

thread 1:
1. deferencing self
2. read barrier
3. increment ref_count
4. (no barrier)

thread 2:
1. deferencing self
2. read barrier
3. read ref_count
4. (still the old one)

Reply:
IMO memory barrier is not a very high cost operation (espessically in x86), if we can not 100% sure about the semantics across the cores and it's not a critical path, selecting always the strongest ordering is an OK solution.

Reply:
Because reordering an operation to the internal `Arc` and operation on this atomic will never result in an unexpected state?

Reply:
reordering includes the compiler reordering, and also the memory reordering.

I guess the Arc dereference operation can prevent the compiler reordering, but  deferencing Arc only produces a read barrier at most, however make a synced modification to other cores requires a write barrier 🤔

Reply:
Right, but I do not think without a write barrier there will be an unexpected result.

The correctness of this piece of code does not depend on the state inside the `Arc`. Thus no matter whether a write operation on another core happened before or after updating this atomic ref_count, it always gets the right result.